### PR TITLE
Use TransactionFailureReason as semantic validation error

### LINKED ISSUE #1565: Refactor semantic errors
- [x] Some variants of `TransactionFailureReason` should be remapped to `StateTransitionError`
- [x] Do we need both ?
- [x] Do we need to respect the TIP for the client facing error ?
- [x] https://github.com/iotaledger/iota-sdk/pull/1572#pullrequestreview-1757551310

https://github.com/iotaledger/iota-core/blob/develop/pkg/retainer/retainer/errors.go

### LINKED ISSUE #2002: Add TxFailureReason ErrInvalidChainAddressUnlock and ErrInvalidDirectUnlockableAddressUnlock
- [x] Would it be possible to have an amount overflow error like the mana one? Could happen with private tangle configurations.
- [x] What error should be used on invalid unlock method? Like a Signature for an AccountAddress?
- [x] What error should be used if an NFT/Account/Anchor unlock point to a different but same type NFT/Account/Anchor?

### PR DESCRIPTION
**Pull request by thibault-martinez**

### CODE REVIEW

**Comment by DaughterOfMars** [comment]
FYI this is also done by #2046

File: bindings/core/src/method_handler/utils.rs [hunk 1]
```diff
@@ -129,7 +129,11 @@ pub(crate) fn call_utils_method_internal(method: UtilsMethod) -> Result<Response
                 protocol_parameters,
             );
 
-            Response::TransactionFailureReason(context.validate()?)
+            if let Err(e) = context.validate() {
+                Response::TransactionFailureReason(Some(e))
+            } else {
+                Response::TransactionFailureReason(None)
+            }
```
Comment (Thoralf-M) [inline bindings/core/src/method_handler/utils.rs]: Don't we now want to return an error in the bindings as well?

Comment (thibault-martinez) [reply bindings/core/src/method_handler/utils.rs]: So removing ` Response::TransactionFailureReason` and letting `validate` error?

Comment (Thoralf-M) [reply bindings/core/src/method_handler/utils.rs]: Yes, but not sure if that's really better, that's why I made it a question

Comment (thibault-martinez) [reply bindings/core/src/method_handler/utils.rs]: Now that it's an actual error, maybe it should behave like one and not return an option yeah

Comment (thibault-martinez) [reply bindings/core/src/method_handler/utils.rs]: I think it's slightly better actually

Comment (thibault-martinez) [reply bindings/core/src/method_handler/utils.rs]: Done

File: sdk/src/client/secret/mod.rs [hunk 2]
```diff
@@ -671,9 +671,9 @@ where
...
     validate_signed_transaction_payload_length(&tx_payload)?;
 
-    let conflict = verify_semantic(&inputs_data, &tx_payload, mana_rewards, protocol_parameters.clone())?;
+    let conflict = verify_semantic(&inputs_data, &tx_payload, mana_rewards, protocol_parameters.clone());
 
-    if let Some(conflict) = conflict {
+    if let Err(conflict) = conflict {
         log::debug!("[sign_transaction] conflict: {conflict:?} for {:#?}", tx_payload);
         return Err(Error::TransactionSemantic(conflict));
     }
```
Comment (Thoralf-M) [inline sdk/src/client/secret/mod.rs]: You can do 
```Rust
    verify_semantic(&inputs_data, &tx_payload, mana_rewards, protocol_parameters.clone())
        .inspect_err(|e| {
            log::debug!("[sign_transaction] conflict: {e:?} for {tx_payload:#?}");
        })
        .map_err(Error::TransactionSemantic)?;
```
copied from https://github.com/iotaledger/iota-sdk/pull/2046 :trollface:

File: sdk/tests/client/signing/account.rs [hunk 3]
```diff
@@ -110,9 +110,9 @@ async fn sign_account_state_transition() -> Result<()> {
         &tx_payload,
         prepared_transaction_data.mana_rewards,
         protocol_parameters,
-    )?;
+    );
```
Comment (Thoralf-M) [inline sdk/tests/client/signing/account.rs]: Just do this and remove the if let Err() below?
```suggestion
    )?;
```

Comment (Thoralf-M) [reply sdk/tests/client/signing/account.rs]: Same in many other tests or with unwrap

Comment (thibault-martinez) [reply sdk/tests/client/signing/account.rs]: I guess it was done this way to include the `tx_payload` but I don't think it's really necessary in tests so yeah I can remove

Comment (thibault-martinez) [reply sdk/tests/client/signing/account.rs]: Done

Comment (Thoralf-M) [reply sdk/tests/client/signing/account.rs]: Yeah usually shouldn't be needed there and if it panics we can add it just to investigate

File: sdk/src/client/secret/mod.rs (hunk 2 above)
Comment (thibault-martinez) [reply sdk/src/client/secret/mod.rs]: stolen

**Review by Thoralf-M — APPROVED** [review]

File: sdk/src/types/block/semantic/mod.rs [hunk 4]
```diff
@@ -145,36 +144,41 @@ impl<'a> SemanticValidationContext<'a> {
...
-                            .ok_or(Error::ConsumedManaOverflow)?;
+                            .checked_add(
+                                consumed_output
+                                    .available_mana(
+                                        &self.protocol_parameters,
+                                        output_id.transaction_id().slot_index(),
+                                        self.transaction.creation_slot(),
+                                    )
+                                    // Unwrap is fine as we already checked both slot indices against each others.
+                                    .unwrap(),
```
Comment (DaughterOfMars) [inline sdk/src/types/block/semantic/mod.rs:163]:
```suggestion
                                    .unwrap(), // Unwrap is fine as we already checked both slot indices against each others.
```

**Review by DaughterOfMars — APPROVED** [review]
We cool with the few instances of `SemanticValidationFailed`?

**Comment by thibault-martinez** [comment]
> We cool with the few instances of `SemanticValidationFailed`?

Yeah it's just failsafe but it should actually never happen

File: sdk/src/types/block/semantic/mod.rs (hunk 4 above)
Comment (thibault-martinez) [reply sdk/src/types/block/semantic/mod.rs:163]: I'll keep it as is if you don't mind, we usually put them on the line before

**Pull request merged**